_posts/2024-04-04-cheriot-supply-chain.markdown

```diff
@@ -0,0 +1,189 @@
+I joined the project in 2012 to lead the compilers / language strand of the research.
+Our first approach involved annotating buffers that may be shared so that they could use CHERI capabilities and everything else could use integer addresses to represent pointers.
+It rapidly became apparent that it was not a good strategy because it meant that all code that touched an object had to be aware that the object might be shared.
+Instead, we moved to the 'pure-capability' model, where every pointer was represented as a capability.
+This involved orders of magnitude less porting effort and had the nice benefit of giving fine-grained memory safety for all objects.
+
+It's worth remembering at this point that memory safety was never the goal of CHERI, it's just something that you need to be able to solve the real problems.
+With object-granularity memory safety, you can build a linkage model that lets you link something like libz or liblzma and know that, at worst, it can tamper with the output buffers that you give it.
+On a CHERI system, every pointer in the source language is lowered to a *CHERI capability*, an unforgeable hardware type that authorises access to memory and carries bounds and permissions.
+You can think of a CHERI capability as a hardware-enforced pointer.
+Every memory access (load, store, or jump) instruction takes a capability that authorises the operation as an operand.
+You can shrink the bounds or remove permissions to create new pointers with restricted rights, and that's a simple register-to-register operation.
+
+This is where the CHERIoT project started.
+We wanted to build a compartmentalisation model (and an associated programmer model) that can rely on CHERI existing and use this to build the kind of supply-chain security where you can fearlessly use third-party code.
+
+Although the rest of this post is specific to CHERIoT, many similar ideas can be applied on bigger CHERI systems such as Arm's Morello.
+The current CheriBSD release includes [support for library compartmentalisation](https://ctsrd-cheri.github.io/cheribsd-getting-started/features/c18n.html) and this is rapidly evolving.
+With CHERIoT, we wanted to showcase what's possible if you can assume CHERI everywhere.
+In contrast, CheriBSD demonstrates how easy it is to incrementally adopt CHERI in environments with strong backwards compatibility requirements.
+
+## The CHERIoT ABI introduces boundaries
+
+UNIX originally did not have any notion of shared libraries.
+When they were added, they replicated the model of static linkage as closely as possible.
+Unlike MULTICS, where libraries could be isolated domains, shared libraries in UNIX all run in the same address space, with the same privileges.
+The liblzma attack took advantage of this to redirect symbols from the main SSH executable into its own functions.
+
+This kind of attack is not possible on the CHERIoT platform.
+The loader runs at startup (just as the run-time linker does on UNIX) but is protected.
```

Comment:
Well, protection is one (true) thing, but it's more that there isn't an analogue of dynamic libraries and especially not IFUNCs in CHERIoT, right?

AIUI, it would, in principle, be possible to generate a compartment object file that exported symbols mangled as if they belonged to a different compartment and thereby cause the (static) linker to have to pick one (or complain), but that would leave traces in its audit log.